File: ruby/lib/move_disks.rb
def move_disks(tower, count, from, to)
  if count == 1
    tower.move(from, to)
  else
    stage = tower.bases.detect { |base| ![from, to].include?(base.id) }.id
    tower = move_disks(tower, count - 1, from, stage)
    tower = move_disks(tower, 1, from, to)
    tower = move_disks(tower, count - 1, stage, to)
    tower
  end
end

File: javascript/lib/moveDisks.js
const moveDisks = (tower, count, from, to) => {
  if (count === 1) {
    return tower.move(from, to);
  } else {
    const stage = tower.bases.find(base => ![from, to].includes(base.id)).id;

    let result = tower;
    result = moveDisks(result, count - 1, from, stage);
    result = moveDisks(result, 1, from, to);
    result = moveDisks(result, count - 1, stage, to);
    return result;
  }
};

export default moveDisks;
